### PR TITLE
fix: do not block drifted nodes from being terminated if consolidation is disabled

### PR DESCRIPTION
**Pull request by saurav-agarwalla**
Fixes #N/A 

**Description**

There was a regression introduced in https://github.com/kubernetes-sigs/karpenter/pull/2180:

```
        // Filter out empty candidates. If there was an empty node that wasn't consolidated before this, we should
        // assume that it was due to budgets. If we don't filter out budgets, users who set a budget for `empty`
        // can find their nodes disrupted here, which while that in itself isn't an issue for empty nodes, it could
        // constrain the `drift` budget.
        if len(candidate.reschedulablePods) == 0 {
            continue
        }
```

This doesn't account for cases where consolidation is disabled (or for that matter has a really long consolidation period). Drift shouldn't be blocked on consolidation since they are independent disruption methods.

So, what happens is that emptiness doesn't disrupt the node since consolidation is disabled. But drift also doesn't terminate the node since the node is empty and it incorrectly assumes that emptiness would always disrupt it.

The original intent makes sense though which is why I am also making a change to prioritize non-empty nodes over empty nodes for drift.

With this change, it is totally possible that nodes which aren't eligible to be consolidated at a moment in time get disrupted due to drift and consume a disruption budget but I think it is more important to still allow drift to happen so that we don't have stale nodes (either with outdated AMI or otherwise) even if they are empty because those empty nodes can still get pods scheduled on them which means that the user runs a risk of having their applications run on outdated AMIs.

This is also why I somehow feel that the original priority order of drift being the first disruption method makes sense since it should be more important for customers to be running a more up-to-date node than being more efficient with their workloads but as the original PR mentions, there are rough edges to that approach especially when customers drift nodes for non-critical reasons.

All of this could've been prevented in the first place if we actually had separate buckets for each disruption reason. Something that https://github.com/kubernetes-sigs/karpenter/issues/2387 intends to solve.

**How was this change tested?**

`make presubmit`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/disruption/drift_test.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ var _ = Describe("Drift", func() {
 		nodePool = test.NodePool(v1.NodePool{
 			Spec: v1.NodePoolSpec{
 				Disruption: v1.Disruption{
-					ConsolidateAfter: v1.MustParseNillableDuration("Never"),
+					ConsolidateAfter: v1.MustParseNillableDuration("1h"),
```
Comment (DerekFrank) [inline pkg/controllers/disruption/drift_test.go:52]: Is there a reason to switch this to 1h?

File: pkg/controllers/disruption/drift_test.go [hunk 2]
```diff
@@ -727,6 +728,48 @@ var _ = Describe("Drift", func() {
 			Expect(ExpectNodes(ctx, env.Client)).To(HaveLen(1))
 			ExpectNotFound(ctx, env.Client, nodeClaim, node)
 		})
+		It("should delete drifted nodes when they are empty and consolidation is disabled", func() {
```
Comment (rschalo) [inline pkg/controllers/disruption/drift_test.go:661]: We should add a test that shows the ordering of disruption given non-empty nodes are being prioritized.

File: pkg/controllers/disruption/drift.go [hunk 3]
```diff
@@ -61,14 +63,15 @@ func (d *Drift) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[
...
-		// assume that it was due to budgets. If we don't filter out budgets, users who set a budget for `empty`
-		// can find their nodes disrupted here, which while that in itself isn't an issue for empty nodes, it could
-		// constrain the `drift` budget.
-		if len(candidate.reschedulablePods) == 0 {
-			continue
-		}
+	emptyCandidates := lo.Filter(candidates, func(c *Candidate, _ int) bool {
+		return len(c.reschedulablePods) == 0
+	})
+	nonEmptyCandidates := lo.Filter(candidates, func(c *Candidate, _ int) bool {
```
Comment (jigisha620) [inline pkg/controllers/disruption/drift.go]: Can we use `lo.FilterReject` for this operation with `len(c.reschedulablePods) == 0` as the predicate. That should give us two slices. One of the slice satisfies the predicate and the other slice doesn't. That way, you don't have to use `lo.Filter` twice.

Comment (saurav-agarwalla) [reply pkg/controllers/disruption/drift.go]: TIL: I'll make the change.

File: pkg/controllers/disruption/drift_test.go (hunk 2 above)
Comment (saurav-agarwalla) [reply pkg/controllers/disruption/drift_test.go:661]: Sure thing. Updating in the next revision.

File: pkg/controllers/disruption/drift_test.go (hunk 1 above)
Comment (saurav-agarwalla) [reply pkg/controllers/disruption/drift_test.go:52]: Just making the default case more representative of real world. It didn't matter earlier since drift was prioritized over consolidation but it does now since emptiness runs first.

File: pkg/controllers/disruption/drift.go [hunk 4]
```diff
@@ -61,14 +63,12 @@ func (d *Drift) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[
...
-		// can find their nodes disrupted here, which while that in itself isn't an issue for empty nodes, it could
-		// constrain the `drift` budget.
-		if len(candidate.reschedulablePods) == 0 {
-			continue
-		}
+	emptyCandidates, nonEmptyCandidates := lo.FilterReject(candidates, func(c *Candidate, _ int) bool {
+		return len(c.reschedulablePods) == 0
+	})
+
+	// Prioritize non-empty candidates since we want them to get priority over empty candidates if the budget is constrained.
```
Comment (saurav-agarwalla) [inline pkg/controllers/disruption/drift.go]: @jmdeal pointed out offline that this might not be desirable since drifting non-empty nodes first might just mean that the pods from those nodes move to the empty nodes leading to more churn. I'll reverse the order in the next revision.

File: pkg/controllers/disruption/drift.go [hunk 5]
```diff
@@ -61,14 +63,12 @@ func (d *Drift) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[
...
-		// can find their nodes disrupted here, which while that in itself isn't an issue for empty nodes, it could
-		// constrain the `drift` budget.
-		if len(candidate.reschedulablePods) == 0 {
-			continue
-		}
+	emptyCandidates, nonEmptyCandidates := lo.FilterReject(candidates, func(c *Candidate, _ int) bool {
+		return len(c.reschedulablePods) == 0
+	})
+
+	// Prioritize empty candidates since we want them to get priority over non-empty candidates if the budget is constrained.
```
Comment (jmdeal) [inline pkg/controllers/disruption/drift.go:70]: Can we update this comment to include why we want to prioritize the non-empty candidates? This is a bit tricky to reason about, so I think it's worth including.

Comment (saurav-agarwalla) [reply pkg/controllers/disruption/drift.go:70]: Done.
